### DISCUSSION TITLE
IF/ELSE question help Micropython programming

### DISCUSSION THREAD

**Discussion by carlstg** [micropython/micropython#9948] [Using MicroPython]
Hello. I'm new to Micropython programming. I'm doing Crestron programming for room automation and this is really different for me with Micropython. But thing progress ;-)

I have an question about interact/stopping a IF / Else programming

I explain more: 
I have setup a Rasbery pico with 3 button and 3 run of 50xWS2811 led  light
If I press button #1 the 1st 50 led light up one after one and stay open for 5 secondes
If I press button #2 the 1st 50 led light up one after one, after done the second 50 led light un one after one and they all stay on 5 sec
If I press button #3 the 1st 50 led light up one after one, after done the second 50 led light un one after one, after done the third 50 led light un one after one and they stay on 5 sec.

Waht I try to achived here and i'm not able to do it. If I press button #1 it start the light up process....but I want stop it and restard an other process if I press button #2 or button #3...

For now the programm work except the ''stop'' function I cannot stop my light to light up by pressing other button.
thanks
Carl

[Map monde v1.22.zip](https://github.com/micropython/micropython/files/10002967/Map.monde.v1.22.zip)

**Comment by jimmo** [comment]
The problem here is that while your code is running the `color_chaseN` function, it's not checking the state of the button pins.

The way to solve this is to either use asyncio, and run the effects in a separate task to the code that checks the pins, or to make the buttons raise interrupts instead.

See https://docs.micropython.org/en/latest/library/machine.Pin.html#machine.Pin.irq

This will allow you to set a function that will be called when the button is pressed, even if it's in the middle of doing the animation or sleeping.

From there you can set a global variable to stop the effect, and check that while you do the animation or sleep.

You will need to rearrange your code a bit, but perhaps the least invasive way to do this is something like. (Untested, but hope you get the idea)

```py
next_button = 0

def button_press(p):
  global next_button
  if p == button1:
    next_button = 1
  #... elif p == button2 (and button3)

# change each of these effect functions to ensure that the most recent button was still the one that started this effect
def color_chase5(color, wait, btn):
    for i in range(NUM_LEDS):
        if next_button != 0 and next_button != btn:        # stop this effect if a new button has been pressed, and it's different to the current one
            return
        pixels_set(i, color)
        time.sleep(wait)
        pixels_show5()

button1.irq(handler=button_press, trigger=Pin.IRQ_FALLING)
button2.irq(handler=button_press, trigger=Pin.IRQ_FALLING)
button3.irq(handler=button_press, trigger=Pin.IRQ_FALLING)

def handle_button1():
    print("country1")
    relay1.off()
    led.value(0)   
    color_chase1(BLUE, 0.2, 1)
    print("country1 ledstrip1 done")
    # 250ms sleep in 25ms chunks
    for i in range(10):
        if next_button != 0 and next_button != 1:
            return
        time.sleep_ms(25)  # <--- fixed indentation
    pixels_fill(BLACK)
    color_chase2(BLUE, 0.2, 1)
    print("country1 ledstrip2 done")
    # 2000ms sleep in 20ms chunks
    for i in range(100):
        if next_button != 0 and next_button != 1:
        return
        time.sleep_ms(20)  # <--- fixed indentation
    relay1.on()      

try:
    while(1):
       current_button = next_button
       next_button = 0
       if current_button == 1:
            handle_button1()
       elif current_button == 2:
          ....
      elif ...
          ....
      else:
         # reset to black
except:
  pass
```

**Reply by carlstg** [reply]
Wow thanks very much I will test taht tonight and let you know my result. Really appreciate your time to answer Jim

**Reply by carlstg** [reply]
hello Jim, I try but there some issue : 
Traceback (most recent call last):
  File "<stdin>", line 230, in <module>
SyntaxError: 'return' outside function

Seem bloked on the ''Return'' inside my IF ELSE button run
I read on some web site how to use the if/els with return / ''short circuit'' my loop
catch better the idea, but not properly ;-)

thanks in advance

[Map monde v1.23.zip](https://github.com/micropython/micropython/files/10008898/Map.monde.v1.23.zip)

Carl

**Reply by jimmo** [reply]
Sorry, I meant to move that code out to a function then forgot. I updated the example.

**Reply by carlstg** [reply]
Thanks Jim, I made some test yesturday night the black at the end don't work anymore but the ''cut'' fonction seem to work. I will try to find waht append in code as much I can tonight before asked help again ;-)

**Reply by carlstg** [reply]
Hello Jim,
thansk for you help and advise 

with version 1.26
The button #2 and #3 can stop any strip in the middle of the process... but not button 1  

I finish to have something working well with version 1.27 button 1 is not used
I jump over button 1 taht don't want to restart it self or cut other made a virtual button#1 
old btn 1 is now 2
old btn 2 is now 3
old btn 3 is now 4
it work but I cannot figure it out why in version 1.26... btn #1 cannot stop other strip in the middle...
prettry sure there is a missign caractere I cannto found ;-) 
;-)

[Map monde v1.26.zip](https://github.com/micropython/micropython/files/10053699/Map.monde.v1.26.zip)
[Map monde v1.27.zip](https://github.com/micropython/micropython/files/10053700/Map.monde.v1.27.zip)

**Reply by jimmo** [reply]
> it work but I cannot figure it out why in version 1.26... btn https://github.com/micropython/micropython/issues/1 cannot stop other strip in the middle...

There are two things:

1. I made a mistake copy&pasting, and the `time.sleep_ms()` need to be _inside_ the for loops. This was preventing the button check from working if the code was in the middle of sleeping. I fixed the example above.
2. When you call `color_chaseN` you need to tell it the current button. So for example in `color_chase1(BLUE, 0.2, 1)`, the `1` is to say this is button 1. You need to change that to `2` in `handle_button2`, etc.

**Reply by carlstg** [reply]
ho yep everithing working well now from 1.26 to 1.28 ;-) Didn't think of the point #2 you reply
I just need to add more to fill all the GPI and add other button for other country
thanks verry much.
Carl
[Map monde v1.28.zip](https://github.com/micropython/micropython/files/10056448/Map.monde.v1.28.zip)
